### DISCUSSION TITLE
Why some plugin opts can be extended with a plain table and some requires a function?

### DISCUSSION THREAD

**Discussion by danielo515** [LazyVim/LazyVim#710] [Q&A] [answered]
I was wondering why some plugins can just get a table of options taht gets merged into the existing one, while others require a function that takes the table and extends them?

```lua
  -- override nvim-cmp and add cmp-emoji
  {
    "hrsh7th/nvim-cmp",
    dependencies = { "hrsh7th/cmp-emoji" },
    ---@param opts cmp.ConfigSchema
    opts = function(_, opts)
      local cmp = require("cmp")
      opts.sources = cmp.config.sources(vim.list_extend(opts.sources, { { name = "emoji" } }))
    end,
  },
```

You may say that it is a matter of preference, but I saw some people reporting specifically they were to able to extend cmp config properly using just a table in place of opts, so I really want to know why is this.

**Comment by folke** [comment, marked as answer]
If you need to require a module to build your opts, you should use a function, otherwise just a table

**Comment by fredrikaverpil** [comment]
@folke I'm also a little confused about when/how to use `opts`.

I have the LazyVim [Rust extras](https://www.lazyvim.org/plugins/extras/lang.rust) enabled, which among other things tell Mason to install `codelldb`. If I have this and run `:Mason`, `codelldb` is _not_ installed:

```lua
return {
  {
    "williamboman/mason.nvim",
    opts = {
      ensure_installed = {
        "rust-analyzer",
        "rustfmt",
      },
    },
  }
}
```

But if I tweak my configuration into this, `codelldb` is _successfully installed_ whenever I run `:Mason`:

```lua
return {
  {
    "williamboman/mason.nvim",
    opts = function(_, opts)
      local ensure_installed = {
        "rust-analyzer",
        "rustfmt",
      }

      -- extend opts.ensure_installed
      for _, package in ipairs(ensure_installed) do
        table.insert(opts.ensure_installed, package)
      end
    end,
  },
```

How come the first code snippet doesn't install `codelldb` if `opts` is supposed to be merged into the parent spec?
Is it because the LazyVim Rust extras is also using `opts` and therefore I am overwriting that in my first example?

**Reply by folke** [reply]
`opts` are merged, but merging of a list is not possible, so merging a list overwrites the list. Check `h: vim.tbl_deep_extend`

**Reply by folke** [reply]
And yes, you were indeed overwriting the `ensure_installed` list
